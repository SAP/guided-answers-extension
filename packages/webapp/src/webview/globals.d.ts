interface Window {
    vscode: {
        postMessage: (message: string | JSON) => void;
    };
}
