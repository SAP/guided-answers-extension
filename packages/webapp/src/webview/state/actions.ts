export {
    updateGuidedAnswerTrees,
    updateActiveNode,
    selectNode,
    goToPreviousPage,
    goToAllAnswers,
    restartAnswer,
    executeCommand,
    setActiveTree,
    searchTree,
    setQueryValue,
    webviewReady,
    getBetaFeatures,
    sendFeedbackOutcome,
    sendFeedbackComment,
    guideFeedback,
    setProductFilters,
    setComponentFilters,
    resetFilters,
    setPageSize,
    feedbackStatus,
    feedbackResponse
} from '@sap/guided-answers-extension-types';
