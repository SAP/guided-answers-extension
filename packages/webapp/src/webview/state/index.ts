export { actions, store } from './store';
export { getInitialState } from './reducers';
