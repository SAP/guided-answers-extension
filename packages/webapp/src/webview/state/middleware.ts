import i18next from 'i18next';
import type { Middleware, MiddlewareAPI, Dispatch, Action } from 'redux';
import type { GuidedAnswerActions } from '@sap/guided-answers-extension-types';
import type { AppState } from '../types';

declare let window: Window;
declare let acquireVsCodeApi: () => typeof window['vscode'];

/**
 * Communication between IDE extension and web view is realized through the communication middleware
 * window event listener for incoming messages in to the webview
 * post message for outgoing messages to the host.
 *
 * @param store - redux store
 * @returns - redux middleware
 */
export const communicationMiddleware: Middleware<
    Dispatch<GuidedAnswerActions>,
    AppState,
    Dispatch<GuidedAnswerActions>
> = (store: MiddlewareAPI) => {
    // Add event handler, this will dispatch incomming state updates
    window.addEventListener('message', (event: MessageEvent) => {
        if (event.origin === window.origin) {
            console.log(i18next.t('MESSAGE_RECEIVED'), event);
            if (event.data && typeof event.data.type === 'string') {
                store.dispatch(event.data);
            }
        } else {
            console.error(i18next.t('UNKNOWN_ORIGIN') + event.origin);
        }
    });
    try {
        window.vscode = acquireVsCodeApi();
    } catch (e) {
        console.error(i18next.t('ERROR_NOT_RUNNING_IN_WEB_VIEW'));
    }
    return (next: Dispatch<GuidedAnswerActions>) =>
        (action): Action => {
            action = next(action);
            if (action && typeof action.type === 'string' && !action.type.startsWith('[view]')) {
                window.vscode.postMessage(action);
                console.log(i18next.t('REACT_ACTION_POSTED'), action);
            }
            return action;
        };
};
