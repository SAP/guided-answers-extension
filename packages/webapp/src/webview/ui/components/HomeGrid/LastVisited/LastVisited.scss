.guided-answer__last-visited {
    width: 100%;
    margin-bottom: 15px;

    >a {
        display: block;
        width: fit-content;
        margin-bottom: 5px;
        color: var(--vscode-foreground);
        font-size: 14px;
        font-weight: 700;
        line-height: 20px;
        text-decoration-line: underline;
        cursor: pointer;

        &:hover {
            text-decoration-line: none;
        }

        &:focus-visible {
            outline: 1px solid var(--vscode-focusBorder, #007fd4);
        }
    }
}
