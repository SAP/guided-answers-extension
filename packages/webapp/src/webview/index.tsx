import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import { App } from './ui/components/App';
import { actions, store } from './state';
import { initI18n } from './i18n';

/**
 * Initialization of i18n, icons
 */
initI18n();

/**
 * Render app
 */
render(
    <Provider store={store}>
        <App />
    </Provider>,
    document.getElementById('root') as HTMLElement
);

/**
 * Fire action which is captured in hosting app (ide-extension) to signal that we are ready to receive messages
 */
actions.webviewReady();

/**Once webview is ready we focus to search field */
document.getElementById('search-field')?.focus();
