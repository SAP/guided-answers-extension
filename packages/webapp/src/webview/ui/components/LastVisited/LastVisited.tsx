import type { ReactElement } from 'react';
import React from 'react';
import { useSelector } from 'react-redux';
import type { AppState } from '../../../types';
import { actions } from '../../../state';
import { FocusZone, FocusZoneDirection } from '@fluentui/react-focus';
import { VscHistory } from 'react-icons/vsc';
import { TreeItemBottomSection } from '../TreeItemBottomSection';
import type { LastVisitedGuides as LastVisitedGuidesType, LastVisitedGuide } from '@sap/guided-answers-extension-types';

/**
 * Shows list of Last Visited Guides.
 *
 * @returns - react elements for LastVisitedGuides view
 */
export function LastVisited(): ReactElement {
    const lastVisitedGuides = useSelector<AppState, LastVisitedGuidesType>((state) => state.lastVisitedGuides);
    if (lastVisitedGuides.length === 0) {
        return <></>;
    }
    return (
        <div>
            <h3 style={{ display: 'flex', alignItems: 'center' }}>
                <VscHistory />
                <span style={{ margin: '0 5px' }}>Last visited</span>
            </h3>
            <FocusZone direction={FocusZoneDirection.bidirectional} isCircularNavigation={true}>
                <ul className="striped-list-items" role="listbox">
                    {lastVisitedGuides.slice(-1).map((lastVisitedGuidesKey: any) => {
                        const lastVisitedGuide: LastVisitedGuide =
                            lastVisitedGuidesKey[Object.keys(lastVisitedGuidesKey)[0]];
                        return (
                            <li
                                key={`tree-item-${lastVisitedGuide.tree.TREE_ID}${lastVisitedGuide.nodePath
                                    .map((n: { NODE_ID: any }) => n.NODE_ID)
                                    .join('-')}`}
                                className="tree-item"
                                role="option">
                                <button
                                    className="guided-answer__tree"
                                    onClick={(): void => {
                                        actions.setActiveTree(lastVisitedGuide.tree);
                                        lastVisitedGuide.nodePath.forEach((node: any) =>
                                            actions.updateActiveNode(node)
                                        );
                                        document.body.focus();
                                    }}>
                                    <div className="guided-answer__tree__ul">
                                        <h3 className="guided-answer__tree__title">
                                            {lastVisitedGuide.tree.TITLE}
                                            {' - '}
                                            {lastVisitedGuide.nodePath[lastVisitedGuide.nodePath.length - 1].TITLE}
                                        </h3>
                                        <TreeItemBottomSection tree={lastVisitedGuide.tree} />
                                    </div>
                                </button>
                            </li>
                        );
                    })}
                </ul>
            </FocusZone>
        </div>
    );
}
