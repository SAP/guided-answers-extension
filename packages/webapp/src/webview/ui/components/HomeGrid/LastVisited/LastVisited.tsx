import type { ReactElement } from 'react';
import React from 'react';
import { useSelector } from 'react-redux';
import type { AppState } from '../../../../types';
import { actions } from '../../../../state';
import { FocusZone, FocusZoneDirection } from '@fluentui/react-focus';
import { UIIcon, UiIcons } from '@sap-ux/ui-components';
import { TreeItemBottomSection } from '../../TreeItemBottomSection';
import type { LastVisitedGuide } from '@sap/guided-answers-extension-types';
import './LastVisited.scss';

/**
 * Shows list of Last Visited Guides.
 *
 * @returns - react elements for LastVisited view
 */
export function LastVisited(): ReactElement {
    const lastVisitedGuides = useSelector<AppState, LastVisitedGuide[]>((state) => state.lastVisitedGuides);
    return (
        <div>
            <h3 className="guided-answer__home-grid__section__title">
                <UIIcon iconName={UiIcons.History} />
                <span>Last visited</span>
            </h3>
            <FocusZone direction={FocusZoneDirection.vertical} isCircularNavigation={true}>
                <ul className="guided-answer__home-grid__section__list" role="listbox">
                    {lastVisitedGuides.slice(-1).map((guide: LastVisitedGuide) => {
                        const guideTitle =
                            guide.nodePath.length > 1
                                ? `${guide.tree.TITLE} - ${guide.nodePath[guide.nodePath.length - 1].TITLE}`
                                : guide.tree.TITLE;
                        return (
                            <li key={`tree-item-${guideTitle}`} className="guided-answer__last-visited">
                                <a
                                    id="last-visited-link"
                                    onClick={(): void => {
                                        actions.setActiveTree(guide.tree);
                                        guide.nodePath.forEach((node: any) => actions.updateActiveNode(node));
                                        document.body.focus();
                                    }}>
                                    {guideTitle}
                                </a>
                                <TreeItemBottomSection product={guide.tree.PRODUCT} component={guide.tree.COMPONENT} />
                            </li>
                        );
                    })}
                </ul>
            </FocusZone>
        </div>
    );
}
