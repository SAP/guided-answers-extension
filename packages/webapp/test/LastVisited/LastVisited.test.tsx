import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { fireEvent, screen } from '@testing-library/dom';
import { useSelector } from 'react-redux';
import { LastVisited } from '../../src/webview/ui/components/LastVisited';
import { actions } from '../../src/webview/state';

jest.mock('../../src/webview/state', () => ({
    actions: {
        setActiveTree: jest.fn(),
        updateActiveNode: jest.fn()
    }
}));

jest.mock('react-redux', () => ({
    useSelector: jest.fn()
}));

describe('<LastVisited />', () => {
    afterEach(cleanup);

    const mockLastVisited = [
        {
            tree: {
                TREE_ID: '1',
                TITLE: 'Bookmark 1 Title',
                DESCRIPTION: 'Bookmark 1 Description',
                PRODUCT: 'Product 1, Product 2',
                COMPONENT: 'Component 1, Component 2'
            },
            nodePath: [{ NODE_ID: '1', TITLE: 'Node 1' }],
            createdAt: 'time'
        },
        {
            tree: {
                TREE_ID: '2',
                TITLE: 'Bookmark 2 Title',
                DESCRIPTION: 'Bookmark 2 Description',
                PRODUCT: 'Product 1, Product 2',
                COMPONENT: 'Component 1, Component 2'
            },
            nodePath: [{ NODE_ID: '2', TITLE: 'Node 2' }],
            createdAt: 'time2'
        }
    ];

    it('Should render LastVisited component', () => {
        (useSelector as jest.Mock).mockImplementation((selector) => selector({ lastVisitedGuides: mockLastVisited }));
        const { container } = render(<LastVisited />);
        expect(container).toMatchSnapshot();

        const btn = screen.getAllByTestId('last-visited-button')[0];
        fireEvent.click(btn);
        expect(actions.setActiveTree).toBeCalledTimes(1);
        expect(actions.updateActiveNode).toBeCalledTimes(1);
    });

    it('Should render LastVisited component with no data', () => {
        (useSelector as jest.Mock).mockImplementation((selector) => selector({ lastVisitedGuides: [] }));
        const { container } = render(<LastVisited />);
        expect(container).toMatchSnapshot();
    });
});
