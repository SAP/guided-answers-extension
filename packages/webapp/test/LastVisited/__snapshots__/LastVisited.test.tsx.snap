// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LastVisited /> Should render LastVisited component 1`] = `
<div>
  <div>
    <h3
      style="display: flex; align-items: center;"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M13.507 12.324a7 7 0 0 0 .065-8.56A7 7 0 0 0 2 4.393V2H1v3.5l.5.5H5V5H2.811a6.008 6.008 0 1 1-.135 5.77l-.887.462a7 7 0 0 0 11.718 1.092zm-3.361-.97l.708-.707L8 7.792V4H7v4l.146.354 3 3z"
          fill-rule="evenodd"
        />
      </svg>
      <span
        style="margin: 0px 5px;"
      >
        Last visited
      </span>
    </h3>
    <div
      class="ms-FocusZone css-101"
      data-focuszone-id="FocusZone0"
    >
      <ul
        class="striped-list-items"
        role="listbox"
      >
        <li
          class="tree-item"
          role="option"
        >
          <button
            class="guided-answer__tree"
            id="last-visited-button"
            tabindex="0"
          >
            <div
              class="guided-answer__tree__ul"
            >
              <h3
                class="guided-answer__tree__title"
              >
                Bookmark 2 Title
                 - 
                Node 2
              </h3>
              <div
                class="bottom-section"
                id="bottom-section"
              >
                <span
                  class="guided-answer__tree__desc"
                >
                  Bookmark 2 Description
                </span>
                <div
                  class="component-and-product-container"
                  style="margin-top: 10px;"
                >
                  <div
                    class="guided-answer__tree__product"
                    id="product-container"
                  >
                    <span
                      class="bottom-title"
                    >
                      Product: 
                    </span>
                    Product 1
                  </div>
                  <div
                    class="guided-answer__tree__component"
                    id="component-container"
                  >
                    <span
                      class="bottom-title"
                    >
                      Component: 
                    </span>
                    Component 1
                  </div>
                </div>
              </div>
            </div>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<LastVisited /> Should render LastVisited component with no data 1`] = `
<div>
  <div>
    <h3
      style="display: flex; align-items: center;"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M13.507 12.324a7 7 0 0 0 .065-8.56A7 7 0 0 0 2 4.393V2H1v3.5l.5.5H5V5H2.811a6.008 6.008 0 1 1-.135 5.77l-.887.462a7 7 0 0 0 11.718 1.092zm-3.361-.97l.708-.707L8 7.792V4H7v4l.146.354 3 3z"
          fill-rule="evenodd"
        />
      </svg>
      <span
        style="margin: 0px 5px;"
      >
        Last visited
      </span>
    </h3>
    <div
      class="ms-FocusZone css-101"
      data-focuszone-id="FocusZone1"
    >
      <ul
        class="striped-list-items"
        role="listbox"
      />
    </div>
  </div>
</div>
`;
